dev/managedservicesplatform/operationdocs/diagram/diagram.go

```diff
@@ -0,0 +1,268 @@
+
+	// Compile an empty input to get an empty graph
+	_, graph, err := d2lib.Compile(context.Background(), "", nil, nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create new diagram")
+	}
+
+	d.graph = graph
+
+	return d, nil
+}
+
+// Generate a diagram from the spec for an environment
+func (d *diagram) Generate(s *spec.Spec, e string) error {
+	// The docs for d2oracle warn that all d2oracle functions are pure: they do not modify the original graph
+	// Therefore when chaining calls you must use the resulting graph from the previous call
+	// https://d2lang.com/tour/api/
+	//
+	// From my testing though something like:
+	// _, _, _ = d2oracle.Create(graph, nil, "a")
+	// _, _, _ = d2oracle.Create(graph, nil, "b")
+	// _, _, _ = d2oracle.Create(graph, nil, "x -> y")
+	//
+	// returns the expected output of `a; b; x -> y`
+	// To be safe I'll use the library as if it is pure
+	graph := d.graph
+
+	env := s.GetEnvironment(e)
+
+	graph, cloudrun, err := cloudrun(graph, env)
+	if err != nil {
+		return errors.Wrap(err, "failed to generate cloudrun")
+	}
+
+	graph, sentry, err := sentry(graph, env)
+	if err != nil {
+		return errors.Wrap(err, "failed to generate sentry")
+	}
+	graph, err = addDirectedConnection(graph, cloudrun, sentry, "")
+	if err != nil {
+		return errors.Wrap(err, "failed to add connection from cloudrun to sentry")
+	}
+
+	graph, monitoring, err := monitoring(graph, env)
+	if err != nil {
+		return errors.Wrap(err, "failed to generate monitoring")
+	}
+	graph, err = addDirectedConnection(graph, cloudrun, monitoring, "")
+	if err != nil {
+		return errors.Wrap(err, "failed to add connection from cloudrun to monitoring")
+	}
+
+	if env.Category != spec.EnvironmentCategoryTest && env.Alerting != nil && pointers.DerefZero(env.Alerting.Opsgenie) {
+		ograph, opsgenie, err := opsgenie(graph, env)
+		if err != nil {
+			return errors.Wrap(err, "failed to generate opsgenie")
+		}
+		ograph, err = addDirectedConnection(ograph, monitoring, opsgenie, "")
+		if err != nil {
+			return errors.Wrap(err, "failed to add connection from monitoring to opsgenie")
+		}
+		graph = ograph
+	}
+
+	if env.EnvironmentServiceSpec != nil {
+		tgraph, trace, err := trace(graph, env)
+		if err != nil {
+			return errors.Wrap(err, "failed to generate trace")
+		}
+		tgraph, err = addDirectedConnection(tgraph, cloudrun, trace, "")
+		if err != nil {
+			return errors.Wrap(err, "failed to add a connection from cloudrun to trace")
+		}
+		graph = tgraph
+	}
+
+	if env.Resources != nil && env.Resources.Redis != nil {
+		rgraph, redis, err := redis(graph, env)
+		if err != nil {
+			return errors.Wrap(err, "failed to generate redis")
+		}
+
+		rgraph, err = addDirectedConnection(rgraph, cloudrun, redis, "")
+		if err != nil {
+			return errors.Wrap(err, "failed to add connection from cloudrun to redis")
+		}
+		graph = rgraph
+	}
+
+	if env.Resources != nil && env.Resources.BigQueryDataset != nil {
+		bgraph, bigquery, err := bigquery(graph, env)
```

Comment:
nit - can we add a verb to all the constructors? for example:

```suggestion
		bgraph, bigquery, err := newBigQueryNode(graph, env)
```

Here, the var `bigquery` will shadow the function `bigquery`, and with the large amount of code here, it gets difficult to read